{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "import os\n",
    "import ipdb\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: ''",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/Users/bzaruk/Documents/Projects/bzaruk/SPECS/notebooks/pipeline_1.ipynb Cell 2'\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/bzaruk/Documents/Projects/bzaruk/SPECS/notebooks/pipeline_1.ipynb#ch0000001?line=0'>1</a>\u001b[0m config_file_path \u001b[39m=\u001b[39m \u001b[39minput\u001b[39m()\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/bzaruk/Documents/Projects/bzaruk/SPECS/notebooks/pipeline_1.ipynb#ch0000001?line=1'>2</a>\u001b[0m config_file \u001b[39m=\u001b[39m \u001b[39mopen\u001b[39;49m(config_file_path)\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/bzaruk/Documents/Projects/bzaruk/SPECS/notebooks/pipeline_1.ipynb#ch0000001?line=2'>3</a>\u001b[0m config \u001b[39m=\u001b[39m json\u001b[39m.\u001b[39mload(config_file)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: ''"
     ]
    }
   ],
   "source": [
    "config_file_path = input()\n",
    "config_file = open(config_file_path)\n",
    "config = json.load(config_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_folder = os.path.join(config['output_folder'], config['experiment_name'])\n",
    "if not os.path.exists(output_folder):\n",
    "    os.makedirs(output_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_df_to_csv(df, output_file):\n",
    "    # create the folder if not exists yet\n",
    "    if not os.path.exists(os.path.dirname(output_file)):\n",
    "        os.makedirs(os.path.dirname(output_file))\n",
    "\n",
    "    # Write DataFrame to csv file\n",
    "    df.to_csv(output_file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def replicate_coverage(lib, df, on=None, on_lib=None, on_df=None):\n",
    "    if on is not None:\n",
    "        lib_len = len(lib[on])\n",
    "        df_len = len(df[on])\n",
    "    elif on_lib is not None and on_df is not None:\n",
    "        lib_len = len(lib[on_lib])\n",
    "        df_len = len(df[on_df])\n",
    "\n",
    "    if lib_len is None or df_len is None:\n",
    "        return None\n",
    "\n",
    "    return df_len/lib_len"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_raw_pairs_file(df):\n",
    "    # drop the id_ columns\n",
    "    # and removed empty rows with unique_17 or barcode\n",
    "    cleaned_df = df.drop(columns=['id_']).dropna(axis=0, subset=['unique_17', 'barcode'])\n",
    "\n",
    "    return cleaned_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_raw_pairs_file_2(df):\n",
    "    # drop the id_ columns\n",
    "    # and removed empty rows with unique_17 or barcode\n",
    "    cleaned_df = df.drop(columns=['id_']).dropna(axis=0, subset=['unique_17'])\n",
    "\n",
    "    return cleaned_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def unique_17_count(df):\n",
    "    unique_17_counts_df = (\n",
    "        df[\"unique_17\"]\n",
    "        .value_counts(sort=True)\n",
    "        .rename_axis(\"unique_17\")\n",
    "        .reset_index(name=\"count\")\n",
    "    )\n",
    "\n",
    "    return unique_17_counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def unique_17_and_bc_count(df):\n",
    "    unique_17__and_bc_counts_df = (\n",
    "        df.value_counts(sort=True)\n",
    "        .rename_axis([\"unique_17\", \"barcode\"])\n",
    "        .reset_index(name=\"count\")\n",
    "    )\n",
    "\n",
    "    return unique_17__and_bc_counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reads_per_million_norm(df, columns, scalling):\n",
    "    norm_df = df.copy()\n",
    "    for c in columns:\n",
    "        total_read_num = df[c].sum()\n",
    "        scalling_factor = total_read_num / scalling\n",
    "        norm_df[c] = norm_df[c] / scalling_factor\n",
    "\n",
    "    return norm_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def num_of_read_by_fastq(fastq_file):\n",
    "    with open(fastq_file, 'r') as f:\n",
    "        x = len(f.readlines())\n",
    "        \n",
    "    # 4 lines per read\n",
    "    return int(x/4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_experimental_dfs(config):\n",
    "    exp_obj = config.copy()\n",
    "\n",
    "    for group in exp_obj[\"groups\"]:\n",
    "\n",
    "        for bio_rep in group[\"biological_replicates\"]:\n",
    "            \n",
    "            for tech_rep in bio_rep[\"technical_replicates\"]:\n",
    "                paired_file = tech_rep[\"paired_file\"]\n",
    "\n",
    "                df = pd.read_csv(paired_file)\n",
    "                cleaned_df = clean_raw_pairs_file(df)\n",
    "                tech_rep[\"df\"] = cleaned_df\n",
    "\n",
    "    return exp_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lib_csv_path = config[\"lib_csv_path\"]\n",
    "lib_df = pd.read_csv(lib_csv_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_experiment_coverage(exp_obj):\n",
    "    coverage_file = os.path.join(output_folder, \"coverage.txt\")\n",
    "    for group in exp_obj[\"groups\"]:\n",
    "        group_name = group[\"name\"]\n",
    "        for bio_rep in group[\"biological_replicates\"]:\n",
    "            bio_name = bio_rep[\"name\"]\n",
    "            for tech_rep in bio_rep[\"technical_replicates\"]:\n",
    "                tech_name = tech_rep[\"name\"]\n",
    "\n",
    "                fastq_file = tech_rep[\"fastq_file\"]\n",
    "                clipped_fastq_file = tech_rep[\"clipped_fastq_file\"]\n",
    "                unclipped_fastq_file = tech_rep[\"unclipped_fastq_file\"]\n",
    "                too_long_fastq_file = tech_rep[\"too_long_fastq_file\"]\n",
    "                too_short_fastq_file = tech_rep[\"too_short_fastq_file\"]\n",
    "\n",
    "                unique_17_count_df = unique_17_count(tech_rep[\"df\"])\n",
    "\n",
    "                coverage = replicate_coverage(\n",
    "                    lib_df, unique_17_count_df, on_lib=\"unique_17\", on_df=\"unique_17\"\n",
    "                )\n",
    "                fastq_reads = num_of_read_by_fastq(fastq_file)\n",
    "                clipped_fastq_reads = num_of_read_by_fastq(clipped_fastq_file)\n",
    "                unclipped_fastq_reads = num_of_read_by_fastq(unclipped_fastq_file)\n",
    "                too_long_fastq_reads = num_of_read_by_fastq(too_long_fastq_file)\n",
    "                too_short_fastq_reads = num_of_read_by_fastq(too_short_fastq_file)\n",
    "\n",
    "                output_line = (\n",
    "                    f\"{group_name} -> {bio_name} -> {tech_name}:\"\n",
    "                    f\"\\n\\tnumber of aligned reads: {tech_rep['df'].shape[0]} ===> coverage: {coverage}\"\n",
    "                    f\"\\n\\tfastq number of reads: {fastq_reads}\"\n",
    "                    f\"\\n\\tclipped fastq number of reads: {clipped_fastq_reads} ===> clipped reads coverage: {clipped_fastq_reads/fastq_reads}\"\n",
    "                    f\"\\n\\tunclipped fastq number of reads: {unclipped_fastq_reads} ===> unclipped reads coverage: {unclipped_fastq_reads/fastq_reads}\"\n",
    "                    f\"\\n\\ttoo long fastq number of reads: {too_long_fastq_reads} ===> too long reads coverage: {too_long_fastq_reads/fastq_reads}\"\n",
    "                    f\"\\n\\ttoo short fastq number of reads: {too_short_fastq_reads} ===> too short reads coverage: {too_short_fastq_reads/fastq_reads}\"\n",
    "                )\n",
    "\n",
    "                print(output_line)\n",
    "                with open(coverage_file, \"a+\") as out:\n",
    "                    out.write(output_line + \"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "expr_obj = read_experimental_dfs(config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H7_R1 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 525534 ===> coverage: 0.606\n",
      "\tfastq number of reads: 2359908\n",
      "\tclipped fastq number of reads: 531876 ===> clipped reads coverage: 0.2253799724396036\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 334810 ===> too long reads coverage: 0.14187417475596506\n",
      "\ttoo short fastq number of reads: 1493222 ===> too short reads coverage: 0.6327458528044314\n",
      "H7_R1 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 41911 ===> coverage: 0.34\n",
      "\tfastq number of reads: 1724697\n",
      "\tclipped fastq number of reads: 42629 ===> clipped reads coverage: 0.024716805328704114\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 239806 ===> too long reads coverage: 0.13904239411328484\n",
      "\ttoo short fastq number of reads: 1442262 ===> too short reads coverage: 0.8362408005580111\n",
      "RL_R1 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 1540268 ===> coverage: 0.9031666666666667\n",
      "\tfastq number of reads: 2269492\n",
      "\tclipped fastq number of reads: 1557410 ===> clipped reads coverage: 0.6862372724821237\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 188184 ===> too long reads coverage: 0.08291899685039647\n",
      "\ttoo short fastq number of reads: 523898 ===> too short reads coverage: 0.23084373066747976\n",
      "RL_R1 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 611366 ===> coverage: 0.7838333333333334\n",
      "\tfastq number of reads: 1940831\n",
      "\tclipped fastq number of reads: 617513 ===> clipped reads coverage: 0.31816938208427215\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 243631 ===> too long reads coverage: 0.12552921918497797\n",
      "\ttoo short fastq number of reads: 1079687 ===> too short reads coverage: 0.5563013987307499\n",
      "DS_R1 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 1709148 ===> coverage: 0.9486666666666667\n",
      "\tfastq number of reads: 2437242\n",
      "\tclipped fastq number of reads: 1727153 ===> clipped reads coverage: 0.7086505976837754\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 189883 ===> too long reads coverage: 0.07790896431294062\n",
      "\ttoo short fastq number of reads: 520206 ===> too short reads coverage: 0.21344043800328405\n",
      "DS_R1 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 740890 ===> coverage: 0.868\n",
      "\tfastq number of reads: 1910958\n",
      "\tclipped fastq number of reads: 747882 ===> clipped reads coverage: 0.3913649593554646\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 215453 ===> too long reads coverage: 0.11274606767914314\n",
      "\ttoo short fastq number of reads: 947623 ===> too short reads coverage: 0.49588897296539225\n",
      "U2_R1 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 63935 ===> coverage: 0.22183333333333333\n",
      "\tfastq number of reads: 2463970\n",
      "\tclipped fastq number of reads: 64831 ===> clipped reads coverage: 0.026311602819839526\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 353742 ===> too long reads coverage: 0.1435658713377192\n",
      "\ttoo short fastq number of reads: 2045397 ===> too short reads coverage: 0.8301225258424413\n",
      "U2_R1 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 8276 ===> coverage: 0.11416666666666667\n",
      "\tfastq number of reads: 1841441\n",
      "\tclipped fastq number of reads: 9014 ===> clipped reads coverage: 0.004895079451364448\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 265024 ===> too long reads coverage: 0.14392206972691496\n",
      "\ttoo short fastq number of reads: 1567403 ===> too short reads coverage: 0.8511828508217206\n",
      "MC_CAR_R1 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 602401 ===> coverage: 0.9203333333333333\n",
      "\tfastq number of reads: 844862\n",
      "\tclipped fastq number of reads: 607769 ===> clipped reads coverage: 0.7193707374695513\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 59823 ===> too long reads coverage: 0.07080801361642493\n",
      "\ttoo short fastq number of reads: 177270 ===> too short reads coverage: 0.20982124891402382\n",
      "MC_CAR_R1 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 635836 ===> coverage: 0.8288333333333333\n",
      "\tfastq number of reads: 1953417\n",
      "\tclipped fastq number of reads: 642135 ===> clipped reads coverage: 0.32872397445092366\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 225637 ===> too long reads coverage: 0.1155088749611578\n",
      "\ttoo short fastq number of reads: 1085645 ===> too short reads coverage: 0.5557671505879185\n",
      "NCI_R1 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 1934379 ===> coverage: 0.9913333333333333\n",
      "\tfastq number of reads: 2007877\n",
      "\tclipped fastq number of reads: 1953769 ===> clipped reads coverage: 0.9730521341695731\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 33296 ===> too long reads coverage: 0.016582689079062112\n",
      "\ttoo short fastq number of reads: 20812 ===> too short reads coverage: 0.01036517675136475\n",
      "NCI_R1 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 1628924 ===> coverage: 0.985\n",
      "\tfastq number of reads: 1793724\n",
      "\tclipped fastq number of reads: 1646259 ===> clipped reads coverage: 0.9177883553991584\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 65399 ===> too long reads coverage: 0.03645990130031153\n",
      "\ttoo short fastq number of reads: 82066 ===> too short reads coverage: 0.04575174330053007\n",
      "HT_1080_R1 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 2169389 ===> coverage: 0.9811666666666666\n",
      "\tfastq number of reads: 2262633\n",
      "\tclipped fastq number of reads: 2191664 ===> clipped reads coverage: 0.9686343300040263\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 62507 ===> too long reads coverage: 0.02762577934645168\n",
      "\ttoo short fastq number of reads: 8462 ===> too short reads coverage: 0.003739890649522039\n",
      "HT_1080_R1 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 1729873 ===> coverage: 0.9301666666666667\n",
      "\tfastq number of reads: 1852713\n",
      "\tclipped fastq number of reads: 1749192 ===> clipped reads coverage: 0.9441246431584385\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 89162 ===> too long reads coverage: 0.04812510086559548\n",
      "\ttoo short fastq number of reads: 14359 ===> too short reads coverage: 0.007750255975966056\n",
      "CD4_R1 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 2068388 ===> coverage: 0.8126666666666666\n",
      "\tfastq number of reads: 2266735\n",
      "\tclipped fastq number of reads: 2090242 ===> clipped reads coverage: 0.922137788493141\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 155511 ===> too long reads coverage: 0.06860572585679402\n",
      "\ttoo short fastq number of reads: 20982 ===> too short reads coverage: 0.009256485650064962\n",
      "CD4_R1 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 1013781 ===> coverage: 0.657\n",
      "\tfastq number of reads: 1288414\n",
      "\tclipped fastq number of reads: 1023252 ===> clipped reads coverage: 0.7941950335839256\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 176154 ===> too long reads coverage: 0.1367215817276124\n",
      "\ttoo short fastq number of reads: 89008 ===> too short reads coverage: 0.06908338468846194\n",
      "CD8_R1 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 2077510 ===> coverage: 0.8433333333333334\n",
      "\tfastq number of reads: 2258303\n",
      "\tclipped fastq number of reads: 2101923 ===> clipped reads coverage: 0.9307533134393392\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 145315 ===> too long reads coverage: 0.06434698975292509\n",
      "\ttoo short fastq number of reads: 11065 ===> too short reads coverage: 0.00489969680773572\n",
      "CD8_R1 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 857348 ===> coverage: 0.7391666666666666\n",
      "\tfastq number of reads: 1075422\n",
      "\tclipped fastq number of reads: 864729 ===> clipped reads coverage: 0.8040834202759475\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 144535 ===> too long reads coverage: 0.13439840360342264\n",
      "\ttoo short fastq number of reads: 66158 ===> too short reads coverage: 0.061518176120629854\n",
      "TREGS_R1 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 2163345 ===> coverage: 0.9813333333333333\n",
      "\tfastq number of reads: 2316060\n",
      "\tclipped fastq number of reads: 2184196 ===> clipped reads coverage: 0.9430653782717201\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 119059 ===> too long reads coverage: 0.05140583577282108\n",
      "\ttoo short fastq number of reads: 12805 ===> too short reads coverage: 0.005528785955458839\n",
      "TREGS_R1 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 1608166 ===> coverage: 0.9781666666666666\n",
      "\tfastq number of reads: 1766254\n",
      "\tclipped fastq number of reads: 1621800 ===> clipped reads coverage: 0.9182144810429304\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 127107 ===> too long reads coverage: 0.0719641682340139\n",
      "\ttoo short fastq number of reads: 17347 ===> too short reads coverage: 0.009821350723055688\n",
      "B_CELLS_R1 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 1403660 ===> coverage: 0.981\n",
      "\tfastq number of reads: 1474655\n",
      "\tclipped fastq number of reads: 1414537 ===> clipped reads coverage: 0.9592324984487898\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 53356 ===> too long reads coverage: 0.03618202223570937\n",
      "\ttoo short fastq number of reads: 6762 ===> too short reads coverage: 0.0045854793155009135\n",
      "B_CELLS_R1 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 1637991 ===> coverage: 0.9605\n",
      "\tfastq number of reads: 1879973\n",
      "\tclipped fastq number of reads: 1655851 ===> clipped reads coverage: 0.8807844580746639\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 166942 ===> too long reads coverage: 0.0888002114923991\n",
      "\ttoo short fastq number of reads: 57180 ===> too short reads coverage: 0.03041533043293707\n",
      "MONOCYTES_R1 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 1649870 ===> coverage: 0.7163333333333334\n",
      "\tfastq number of reads: 1949403\n",
      "\tclipped fastq number of reads: 1664904 ===> clipped reads coverage: 0.8540583963398025\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 232962 ===> too long reads coverage: 0.11950427900234072\n",
      "\ttoo short fastq number of reads: 51537 ===> too short reads coverage: 0.02643732465785679\n",
      "MONOCYTES_R1 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 429681 ===> coverage: 0.6091666666666666\n",
      "\tfastq number of reads: 724626\n",
      "\tclipped fastq number of reads: 434327 ===> clipped reads coverage: 0.599380922020463\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 182196 ===> too long reads coverage: 0.25143453312467395\n",
      "\ttoo short fastq number of reads: 108103 ===> too short reads coverage: 0.14918454485486307\n",
      "JURKAT_R1 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 1558423 ===> coverage: 0.9846666666666667\n",
      "\tfastq number of reads: 1633646\n",
      "\tclipped fastq number of reads: 1575395 ===> clipped reads coverage: 0.9643429482274618\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 50719 ===> too long reads coverage: 0.031046505791340353\n",
      "\ttoo short fastq number of reads: 7532 ===> too short reads coverage: 0.004610545981197885\n",
      "JURKAT_R1 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 1791229 ===> coverage: 0.969\n",
      "\tfastq number of reads: 1931701\n",
      "\tclipped fastq number of reads: 1811760 ===> clipped reads coverage: 0.937909127758385\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 106553 ===> too long reads coverage: 0.055160193011237243\n",
      "\ttoo short fastq number of reads: 13388 ===> too short reads coverage: 0.006930679230377786\n",
      "H1_R1 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 1520566 ===> coverage: 0.9838333333333333\n",
      "\tfastq number of reads: 1574217\n",
      "\tclipped fastq number of reads: 1535544 ===> clipped reads coverage: 0.975433501226324\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 32606 ===> too long reads coverage: 0.020712519303247266\n",
      "\ttoo short fastq number of reads: 6067 ===> too short reads coverage: 0.003853979470428791\n",
      "H1_R1 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 1891395 ===> coverage: 0.9696666666666667\n",
      "\tfastq number of reads: 2013523\n",
      "\tclipped fastq number of reads: 1912039 ===> clipped reads coverage: 0.9495987877963152\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 89080 ===> too long reads coverage: 0.04424086538867448\n",
      "\ttoo short fastq number of reads: 12404 ===> too short reads coverage: 0.006160346815010308\n",
      "H7_R2 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 502886 ===> coverage: 0.5911666666666666\n",
      "\tfastq number of reads: 3466961\n",
      "\tclipped fastq number of reads: 503561 ===> clipped reads coverage: 0.14524564885500588\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 22351 ===> too long reads coverage: 0.006446856483242817\n",
      "\ttoo short fastq number of reads: 2941049 ===> too short reads coverage: 0.8483074946617513\n",
      "H7_R2 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 91419 ===> coverage: 0.357\n",
      "\tfastq number of reads: 5378127\n",
      "\tclipped fastq number of reads: 92036 ===> clipped reads coverage: 0.017113020945767922\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 30847 ===> too long reads coverage: 0.0057356399356132725\n",
      "\ttoo short fastq number of reads: 5255244 ===> too short reads coverage: 0.9771513391186188\n",
      "RL_R2 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 1350878 ===> coverage: 0.8995\n",
      "\tfastq number of reads: 2059836\n",
      "\tclipped fastq number of reads: 1353683 ===> clipped reads coverage: 0.6571799890865098\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 15582 ===> too long reads coverage: 0.007564679906555668\n",
      "\ttoo short fastq number of reads: 690571 ===> too short reads coverage: 0.33525533100693455\n",
      "RL_R2 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 938707 ===> coverage: 0.7941666666666667\n",
      "\tfastq number of reads: 3646243\n",
      "\tclipped fastq number of reads: 940215 ===> clipped reads coverage: 0.2578585683949205\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 22146 ===> too long reads coverage: 0.006073648958667867\n",
      "\ttoo short fastq number of reads: 2683882 ===> too short reads coverage: 0.7360677826464117\n",
      "DS_R2 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 1131540 ===> coverage: 0.9423333333333334\n",
      "\tfastq number of reads: 1678023\n",
      "\tclipped fastq number of reads: 1133868 ===> clipped reads coverage: 0.6757166022158219\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 12757 ===> too long reads coverage: 0.00760239877522537\n",
      "\ttoo short fastq number of reads: 531398 ===> too short reads coverage: 0.3166809990089528\n",
      "DS_R2 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 1068501 ===> coverage: 0.8746666666666667\n",
      "\tfastq number of reads: 3606083\n",
      "\tclipped fastq number of reads: 1071016 ===> clipped reads coverage: 0.2970025925637319\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 23414 ===> too long reads coverage: 0.0064929176616289755\n",
      "\ttoo short fastq number of reads: 2511653 ===> too short reads coverage: 0.6965044897746391\n",
      "U2_R2 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 74816 ===> coverage: 0.19283333333333333\n",
      "\tfastq number of reads: 4838617\n",
      "\tclipped fastq number of reads: 74957 ===> clipped reads coverage: 0.015491410045473738\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 24952 ===> too long reads coverage: 0.005156845437446278\n",
      "\ttoo short fastq number of reads: 4738708 ===> too short reads coverage: 0.97935174451708\n",
      "U2_R2 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 13999 ===> coverage: 0.09666666666666666\n",
      "\tfastq number of reads: 5193018\n",
      "\tclipped fastq number of reads: 14809 ===> clipped reads coverage: 0.0028517135892846897\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 32577 ===> too long reads coverage: 0.006273230710927634\n",
      "\ttoo short fastq number of reads: 5145632 ===> too short reads coverage: 0.9908750556997876\n",
      "MC_CAR_R2 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 1473557 ===> coverage: 0.933\n",
      "\tfastq number of reads: 2098981\n",
      "\tclipped fastq number of reads: 1476644 ===> clipped reads coverage: 0.7035051770359045\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 15905 ===> too long reads coverage: 0.007577486408881262\n",
      "\ttoo short fastq number of reads: 606432 ===> too short reads coverage: 0.2889173365552142\n",
      "MC_CAR_R2 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 884036 ===> coverage: 0.83\n",
      "\tfastq number of reads: 3821931\n",
      "\tclipped fastq number of reads: 885857 ===> clipped reads coverage: 0.23178257273613784\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 24335 ===> too long reads coverage: 0.006367200245111699\n",
      "\ttoo short fastq number of reads: 2911739 ===> too short reads coverage: 0.7618502270187505\n",
      "NCI_R2 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 1702734 ===> coverage: 0.988\n",
      "\tfastq number of reads: 1737705\n",
      "\tclipped fastq number of reads: 1706159 ===> clipped reads coverage: 0.9818461706676335\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 13166 ===> too long reads coverage: 0.007576660019968867\n",
      "\ttoo short fastq number of reads: 18380 ===> too short reads coverage: 0.010577169312397675\n",
      "NCI_R2 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 2021533 ===> coverage: 0.9858333333333333\n",
      "\tfastq number of reads: 2155775\n",
      "\tclipped fastq number of reads: 2025542 ===> clipped reads coverage: 0.9395887789773979\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 14230 ===> too long reads coverage: 0.006600874395519013\n",
      "\ttoo short fastq number of reads: 116003 ===> too short reads coverage: 0.05381034662708307\n",
      "HT_1080_R2 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 1446382 ===> coverage: 0.9743333333333334\n",
      "\tfastq number of reads: 1463607\n",
      "\tclipped fastq number of reads: 1448786 ===> clipped reads coverage: 0.9898736477756666\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 9664 ===> too long reads coverage: 0.006602865386678255\n",
      "\ttoo short fastq number of reads: 5157 ===> too short reads coverage: 0.00352348683765519\n",
      "HT_1080_R2 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 1842050 ===> coverage: 0.93\n",
      "\tfastq number of reads: 2498912\n",
      "\tclipped fastq number of reads: 1845575 ===> clipped reads coverage: 0.7385514175769294\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 10961 ===> too long reads coverage: 0.004386308921642699\n",
      "\ttoo short fastq number of reads: 14494 ===> too short reads coverage: 0.005800124214057958\n",
      "CD4_R2 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 2010158 ===> coverage: 0.8115\n",
      "\tfastq number of reads: 2045234\n",
      "\tclipped fastq number of reads: 2013600 ===> clipped reads coverage: 0.9845328211832973\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 13957 ===> too long reads coverage: 0.0068241580181045295\n",
      "\ttoo short fastq number of reads: 17677 ===> too short reads coverage: 0.008643020798598107\n",
      "CD4_R2 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 4029129 ===> coverage: 0.6768333333333333\n",
      "\tfastq number of reads: 4434493\n",
      "\tclipped fastq number of reads: 4035718 ===> clipped reads coverage: 0.9100742745563022\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 30209 ===> too long reads coverage: 0.006812278201814729\n",
      "\ttoo short fastq number of reads: 368566 ===> too short reads coverage: 0.08311344724188312\n",
      "CD8_R2 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 2267204 ===> coverage: 0.8423333333333334\n",
      "\tfastq number of reads: 2296241\n",
      "\tclipped fastq number of reads: 2271653 ===> clipped reads coverage: 0.9892920647266554\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 14385 ===> too long reads coverage: 0.006264586339151683\n",
      "\ttoo short fastq number of reads: 10203 ===> too short reads coverage: 0.004443348934192883\n",
      "CD8_R2 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 3495990 ===> coverage: 0.7728333333333334\n",
      "\tfastq number of reads: 3783346\n",
      "\tclipped fastq number of reads: 3504515 ===> clipped reads coverage: 0.9263004229589363\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 22551 ===> too long reads coverage: 0.005960596783905041\n",
      "\ttoo short fastq number of reads: 256280 ===> too short reads coverage: 0.0677389802571586\n",
      "TREGS_R2 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 1968151 ===> coverage: 0.9805\n",
      "\tfastq number of reads: 1995249\n",
      "\tclipped fastq number of reads: 1970452 ===> clipped reads coverage: 0.9875719772319144\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 13557 ===> too long reads coverage: 0.0067946406689089935\n",
      "\ttoo short fastq number of reads: 11240 ===> too short reads coverage: 0.005633382099176594\n",
      "TREGS_R2 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 956780 ===> coverage: 0.9741666666666666\n",
      "\tfastq number of reads: 972740\n",
      "\tclipped fastq number of reads: 957935 ===> clipped reads coverage: 0.9847801056808603\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 6406 ===> too long reads coverage: 0.006585521310936118\n",
      "\ttoo short fastq number of reads: 8399 ===> too short reads coverage: 0.008634373008203631\n",
      "B_CELLS_R2 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 1388872 ===> coverage: 0.9803333333333333\n",
      "\tfastq number of reads: 1406249\n",
      "\tclipped fastq number of reads: 1390770 ===> clipped reads coverage: 0.9889927032837001\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 9249 ===> too long reads coverage: 0.006577071343695178\n",
      "\ttoo short fastq number of reads: 6230 ===> too short reads coverage: 0.0044302253726047095\n",
      "B_CELLS_R2 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 3136451 ===> coverage: 0.9655\n",
      "\tfastq number of reads: 3270643\n",
      "\tclipped fastq number of reads: 3139738 ===> clipped reads coverage: 0.9599757601181175\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 18830 ===> too long reads coverage: 0.005757277697382442\n",
      "\ttoo short fastq number of reads: 112075 ===> too short reads coverage: 0.034266962184500115\n",
      "MONOCYTES_R2 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 2598587 ===> coverage: 0.7168333333333333\n",
      "\tfastq number of reads: 2707621\n",
      "\tclipped fastq number of reads: 2602590 ===> clipped reads coverage: 0.9612091204788262\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 22282 ===> too long reads coverage: 0.008229364449455814\n",
      "\ttoo short fastq number of reads: 82749 ===> too short reads coverage: 0.03056151507171794\n",
      "MONOCYTES_R2 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 3110244 ===> coverage: 0.6291666666666667\n",
      "\tfastq number of reads: 3981141\n",
      "\tclipped fastq number of reads: 3114240 ===> clipped reads coverage: 0.7822481042495104\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 46099 ===> too long reads coverage: 0.011579343710760307\n",
      "\ttoo short fastq number of reads: 820802 ===> too short reads coverage: 0.2061725520397293\n",
      "JURKAT_R2 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 1763820 ===> coverage: 0.9861666666666666\n",
      "\tfastq number of reads: 1786022\n",
      "\tclipped fastq number of reads: 1767288 ===> clipped reads coverage: 0.9895107675045436\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 11760 ===> too long reads coverage: 0.006584465364928315\n",
      "\ttoo short fastq number of reads: 6974 ===> too short reads coverage: 0.0039047671305280675\n",
      "JURKAT_R2 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 2549929 ===> coverage: 0.972\n",
      "\tfastq number of reads: 2589632\n",
      "\tclipped fastq number of reads: 2554520 ===> clipped reads coverage: 0.9864413167585201\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 16562 ===> too long reads coverage: 0.006395503299310481\n",
      "\ttoo short fastq number of reads: 18550 ===> too short reads coverage: 0.007163179942169389\n",
      "H1_R2 -> bio_1 -> tech_2:\n",
      "\tnumber of aligned reads: 1466688 ===> coverage: 0.9815\n",
      "\tfastq number of reads: 1486561\n",
      "\tclipped fastq number of reads: 1469584 ===> clipped reads coverage: 0.9885796815603262\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 11352 ===> too long reads coverage: 0.007636417207231994\n",
      "\ttoo short fastq number of reads: 5625 ===> too short reads coverage: 0.0037839012324418575\n",
      "H1_R2 -> bio_1 -> tech_3:\n",
      "\tnumber of aligned reads: 2422314 ===> coverage: 0.9715\n",
      "\tfastq number of reads: 2460424\n",
      "\tclipped fastq number of reads: 2427017 ===> clipped reads coverage: 0.9864222589277295\n",
      "\tunclipped fastq number of reads: 0 ===> unclipped reads coverage: 0.0\n",
      "\ttoo long fastq number of reads: 17609 ===> too long reads coverage: 0.007156896534906179\n",
      "\ttoo short fastq number of reads: 15798 ===> too short reads coverage: 0.006420844537364291\n"
     ]
    }
   ],
   "source": [
    "run_experiment_coverage(expr_obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6000"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_unique_17s = lib_df[\"unique_17\"]\n",
    "len(all_unique_17s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_experiment_analysis(expr_obj):\n",
    "    expand_output_df = pd.DataFrame(data={\"unique_17\": all_unique_17s})\n",
    "    bio_output_df = pd.DataFrame(data={\"unique_17\": all_unique_17s})\n",
    "\n",
    "    for group in expr_obj[\"groups\"]:\n",
    "        group_name = group[\"name\"]\n",
    "        for bio_rep in group[\"biological_replicates\"]:\n",
    "            bio_name = bio_rep[\"name\"]\n",
    "            bio_count_col_name = f\"{group_name}_{bio_name}\"\n",
    "            bio_count_df = pd.DataFrame(data={\"unique_17\": all_unique_17s, f\"{bio_count_col_name}\": np.zeros(len(all_unique_17s))})\n",
    "            for tech_rep in bio_rep[\"technical_replicates\"]:\n",
    "                tech_name = tech_rep[\"name\"]\n",
    "                \n",
    "                unique_17_count_df = unique_17_count(tech_rep[\"df\"])\n",
    "                tech_count_col_name = f\"{group_name}_{bio_name}_{tech_name}\"\n",
    "                unique_17_count_df = unique_17_count_df.rename(columns={\"count\" : tech_count_col_name})\n",
    "                bio_count_df = bio_count_df.merge(unique_17_count_df, how=\"left\", on=\"unique_17\").fillna(0)\n",
    "                bio_count_df[bio_count_col_name] = bio_count_df[bio_count_col_name] + bio_count_df[tech_count_col_name]\n",
    "            \n",
    "            bio_output_df = bio_output_df.merge(bio_count_df[[\"unique_17\",bio_count_col_name]], how=\"left\", on=\"unique_17\")\n",
    "            expand_output_df = expand_output_df.merge(bio_count_df, how=\"left\", on=\"unique_17\")\n",
    "\n",
    "    return expand_output_df, bio_output_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "expand_output_df, bio_output_df = run_experiment_analysis(expr_obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>unique_17</th>\n",
       "      <th>H7_R1_bio_1</th>\n",
       "      <th>RL_R1_bio_1</th>\n",
       "      <th>DS_R1_bio_1</th>\n",
       "      <th>U2_R1_bio_1</th>\n",
       "      <th>MC_CAR_R1_bio_1</th>\n",
       "      <th>NCI_R1_bio_1</th>\n",
       "      <th>HT_1080_R1_bio_1</th>\n",
       "      <th>CD4_R1_bio_1</th>\n",
       "      <th>CD8_R1_bio_1</th>\n",
       "      <th>...</th>\n",
       "      <th>MC_CAR_R2_bio_1</th>\n",
       "      <th>NCI_R2_bio_1</th>\n",
       "      <th>HT_1080_R2_bio_1</th>\n",
       "      <th>CD4_R2_bio_1</th>\n",
       "      <th>CD8_R2_bio_1</th>\n",
       "      <th>TREGS_R2_bio_1</th>\n",
       "      <th>B_CELLS_R2_bio_1</th>\n",
       "      <th>MONOCYTES_R2_bio_1</th>\n",
       "      <th>JURKAT_R2_bio_1</th>\n",
       "      <th>H1_R2_bio_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACTACCTGAAGAACCTT</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>234.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAGCTAAATGGCTGATT</td>\n",
       "      <td>53.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTGACCACACTTACAGT</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>176.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTGTTGGCGAGCAGTGT</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAATATCGGCGAGCTCT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>...</td>\n",
       "      <td>112.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>305.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>172.0</td>\n",
       "      <td>121.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           unique_17  H7_R1_bio_1  RL_R1_bio_1  DS_R1_bio_1  U2_R1_bio_1  \\\n",
       "0  ACTACCTGAAGAACCTT          0.0          1.0          0.0          0.0   \n",
       "1  GAGCTAAATGGCTGATT         53.0          0.0          0.0          0.0   \n",
       "2  GTGACCACACTTACAGT          0.0          0.0          7.0          0.0   \n",
       "3  TTGTTGGCGAGCAGTGT          0.0          0.0          9.0          0.0   \n",
       "4  CAATATCGGCGAGCTCT          1.0         68.0         40.0          0.0   \n",
       "\n",
       "   MC_CAR_R1_bio_1  NCI_R1_bio_1  HT_1080_R1_bio_1  CD4_R1_bio_1  \\\n",
       "0              3.0           2.0               1.0           0.0   \n",
       "1              0.0          45.0               0.0           0.0   \n",
       "2             15.0          23.0              17.0          81.0   \n",
       "3              9.0           4.0               0.0           0.0   \n",
       "4             61.0          63.0              38.0          22.0   \n",
       "\n",
       "   CD8_R1_bio_1  ...  MC_CAR_R2_bio_1  NCI_R2_bio_1  HT_1080_R2_bio_1  \\\n",
       "0           0.0  ...              6.0           0.0               0.0   \n",
       "1          46.0  ...              4.0          60.0               0.0   \n",
       "2          14.0  ...             35.0          23.0              15.0   \n",
       "3           0.0  ...             13.0           2.0               2.0   \n",
       "4          98.0  ...            112.0          66.0              31.0   \n",
       "\n",
       "   CD4_R2_bio_1  CD8_R2_bio_1  TREGS_R2_bio_1  B_CELLS_R2_bio_1  \\\n",
       "0           0.0           0.0             0.0              96.0   \n",
       "1           0.0          98.0             5.0              89.0   \n",
       "2          78.0          57.0            45.0             176.0   \n",
       "3           0.0           0.0             3.0               8.0   \n",
       "4          26.0         305.0            23.0              81.0   \n",
       "\n",
       "   MONOCYTES_R2_bio_1  JURKAT_R2_bio_1  H1_R2_bio_1  \n",
       "0               234.0             17.0          1.0  \n",
       "1                 0.0              4.0          2.0  \n",
       "2                19.0             42.0         45.0  \n",
       "3                 0.0             20.0          0.0  \n",
       "4                 2.0            172.0        121.0  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bio_output_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>unique_17</th>\n",
       "      <th>H7_R1_bio_1</th>\n",
       "      <th>H7_R1_bio_1_tech_2</th>\n",
       "      <th>H7_R1_bio_1_tech_3</th>\n",
       "      <th>RL_R1_bio_1</th>\n",
       "      <th>RL_R1_bio_1_tech_2</th>\n",
       "      <th>RL_R1_bio_1_tech_3</th>\n",
       "      <th>DS_R1_bio_1</th>\n",
       "      <th>DS_R1_bio_1_tech_2</th>\n",
       "      <th>DS_R1_bio_1_tech_3</th>\n",
       "      <th>...</th>\n",
       "      <th>B_CELLS_R2_bio_1_tech_3</th>\n",
       "      <th>MONOCYTES_R2_bio_1</th>\n",
       "      <th>MONOCYTES_R2_bio_1_tech_2</th>\n",
       "      <th>MONOCYTES_R2_bio_1_tech_3</th>\n",
       "      <th>JURKAT_R2_bio_1</th>\n",
       "      <th>JURKAT_R2_bio_1_tech_2</th>\n",
       "      <th>JURKAT_R2_bio_1_tech_3</th>\n",
       "      <th>H1_R2_bio_1</th>\n",
       "      <th>H1_R2_bio_1_tech_2</th>\n",
       "      <th>H1_R2_bio_1_tech_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACTACCTGAAGAACCTT</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>85.0</td>\n",
       "      <td>234.0</td>\n",
       "      <td>234.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAGCTAAATGGCTGATT</td>\n",
       "      <td>53.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>87.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTGACCACACTTACAGT</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>109.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>37.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTGTTGGCGAGCAGTGT</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAATATCGGCGAGCTCT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>...</td>\n",
       "      <td>62.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>172.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>73.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 85 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           unique_17  H7_R1_bio_1  H7_R1_bio_1_tech_2  H7_R1_bio_1_tech_3  \\\n",
       "0  ACTACCTGAAGAACCTT          0.0                 0.0                 0.0   \n",
       "1  GAGCTAAATGGCTGATT         53.0                53.0                 0.0   \n",
       "2  GTGACCACACTTACAGT          0.0                 0.0                 0.0   \n",
       "3  TTGTTGGCGAGCAGTGT          0.0                 0.0                 0.0   \n",
       "4  CAATATCGGCGAGCTCT          1.0                 0.0                 1.0   \n",
       "\n",
       "   RL_R1_bio_1  RL_R1_bio_1_tech_2  RL_R1_bio_1_tech_3  DS_R1_bio_1  \\\n",
       "0          1.0                 0.0                 1.0          0.0   \n",
       "1          0.0                 0.0                 0.0          0.0   \n",
       "2          0.0                 0.0                 0.0          7.0   \n",
       "3          0.0                 0.0                 0.0          9.0   \n",
       "4         68.0                48.0                20.0         40.0   \n",
       "\n",
       "   DS_R1_bio_1_tech_2  DS_R1_bio_1_tech_3  ...  B_CELLS_R2_bio_1_tech_3  \\\n",
       "0                 0.0                 0.0  ...                     85.0   \n",
       "1                 0.0                 0.0  ...                     87.0   \n",
       "2                 7.0                 0.0  ...                    109.0   \n",
       "3                 6.0                 3.0  ...                      2.0   \n",
       "4                28.0                12.0  ...                     62.0   \n",
       "\n",
       "   MONOCYTES_R2_bio_1  MONOCYTES_R2_bio_1_tech_2  MONOCYTES_R2_bio_1_tech_3  \\\n",
       "0               234.0                      234.0                        0.0   \n",
       "1                 0.0                        0.0                        0.0   \n",
       "2                19.0                       19.0                        0.0   \n",
       "3                 0.0                        0.0                        0.0   \n",
       "4                 2.0                        2.0                        0.0   \n",
       "\n",
       "   JURKAT_R2_bio_1  JURKAT_R2_bio_1_tech_2  JURKAT_R2_bio_1_tech_3  \\\n",
       "0             17.0                    13.0                     4.0   \n",
       "1              4.0                     2.0                     2.0   \n",
       "2             42.0                    19.0                    23.0   \n",
       "3             20.0                     7.0                    13.0   \n",
       "4            172.0                    81.0                    91.0   \n",
       "\n",
       "   H1_R2_bio_1  H1_R2_bio_1_tech_2  H1_R2_bio_1_tech_3  \n",
       "0          1.0                 1.0                 0.0  \n",
       "1          2.0                 0.0                 2.0  \n",
       "2         45.0                 8.0                37.0  \n",
       "3          0.0                 0.0                 0.0  \n",
       "4        121.0                48.0                73.0  \n",
       "\n",
       "[5 rows x 85 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expand_output_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "expand_output_df_csv = os.path.join(output_folder, \"all_replicates_sum.csv\")\n",
    "save_df_to_csv(expand_output_df, expand_output_df_csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "bio_output_df_csv = os.path.join(output_folder, \"all_bio_replicates_sum.csv\")\n",
    "save_df_to_csv(bio_output_df, bio_output_df_csv)"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "70b838359764d1d529f23188ff72e6bf01483479c193c40f5e02842d85becc68"
  },
  "kernelspec": {
   "display_name": "Python 3.8.12 ('bio')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
