{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "import os\n",
    "import ipdb\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "config_file_path = input()\n",
    "config_file = open(config_file_path)\n",
    "config = json.load(config_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_folder = os.path.join(config['output_folder'], config['experiment_name'])\n",
    "if not os.path.exists(output_folder):\n",
    "    os.makedirs(output_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_df_to_csv(df, output_file):\n",
    "    # create the folder if not exists yet\n",
    "    if not os.path.exists(os.path.dirname(output_file)):\n",
    "        os.makedirs(os.path.dirname(output_file))\n",
    "\n",
    "    # Write DataFrame to csv file\n",
    "    df.to_csv(output_file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def replicate_coverage(lib, df, on=None, on_lib=None, on_df=None):\n",
    "    if on is not None:\n",
    "        lib_len = len(lib[on])\n",
    "        df_len = len(df[on])\n",
    "    elif on_lib is not None and on_df is not None:\n",
    "        lib_len = len(lib[on_lib])\n",
    "        df_len = len(df[on_df])\n",
    "\n",
    "    if lib_len is None or df_len is None:\n",
    "        return None\n",
    "\n",
    "    return df_len/lib_len"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_raw_pairs_file(df):\n",
    "    # drop the id_ columns\n",
    "    # and removed empty rows with unique_17 or barcode\n",
    "    cleaned_df = df.drop(columns=['id_']).dropna(axis=0, subset=['unique_17', 'barcode'])\n",
    "\n",
    "    return cleaned_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_raw_pairs_file_2(df):\n",
    "    # drop the id_ columns\n",
    "    # and removed empty rows with unique_17 or barcode\n",
    "    cleaned_df = df.drop(columns=['id_']).dropna(axis=0, subset=['unique_17'])\n",
    "\n",
    "    return cleaned_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def unique_17_count(df):\n",
    "    unique_17_counts_df = (\n",
    "        df[\"unique_17\"]\n",
    "        .value_counts(sort=True)\n",
    "        .rename_axis(\"unique_17\")\n",
    "        .reset_index(name=\"count\")\n",
    "    )\n",
    "\n",
    "    return unique_17_counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def unique_17_and_bc_count(df):\n",
    "    unique_17__and_bc_counts_df = (\n",
    "        df.value_counts(sort=True)\n",
    "        .rename_axis([\"unique_17\", \"barcode\"])\n",
    "        .reset_index(name=\"count\")\n",
    "    )\n",
    "\n",
    "    return unique_17__and_bc_counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reads_per_million_norm(df, scalling=1e6):\n",
    "    norm_df = df.copy()\n",
    "    for column in norm_df:\n",
    "        total_read_num = df[column].sum()\n",
    "        scalling_factor = total_read_num / scalling\n",
    "        norm_df[column] = norm_df[column] / scalling_factor\n",
    "\n",
    "    return norm_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def num_of_read_by_fastq(fastq_file):\n",
    "    with open(fastq_file, 'r') as f:\n",
    "        x = len(f.readlines())\n",
    "        \n",
    "    # 4 lines per read\n",
    "    return int(x/4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_experimental_dfs(config):\n",
    "    exp_obj = config.copy()\n",
    "\n",
    "    for group in exp_obj[\"groups\"]:\n",
    "\n",
    "        for bio_rep in group[\"biological_replicates\"]:\n",
    "            \n",
    "            for tech_rep in bio_rep[\"technical_replicates\"]:\n",
    "                paired_file = tech_rep[\"paired_file\"]\n",
    "\n",
    "                df = pd.read_csv(paired_file)\n",
    "                cleaned_df = clean_raw_pairs_file(df)\n",
    "                tech_rep[\"df\"] =  cleaned_df\n",
    "\n",
    "    return exp_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "lib_csv_path = config[\"lib_csv_path\"]\n",
    "lib_df = pd.read_csv(lib_csv_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_experiment_coverage(exp_obj):\n",
    "    coverage_file = os.path.join(output_folder, \"coverage.txt\")\n",
    "    for group in exp_obj[\"groups\"]:\n",
    "        group_name = group[\"name\"]\n",
    "        for bio_rep in group[\"biological_replicates\"]:\n",
    "            bio_name = bio_rep[\"name\"]\n",
    "            for tech_rep in bio_rep[\"technical_replicates\"]:\n",
    "                tech_name = tech_rep[\"name\"]\n",
    "\n",
    "                fastq_file = tech_rep[\"fastq_file\"]\n",
    "                clipped_fastq_file = tech_rep[\"clipped_fastq_file\"]\n",
    "                unclipped_fastq_file = tech_rep[\"unclipped_fastq_file\"]\n",
    "                too_long_fastq_file = tech_rep[\"too_long_fastq_file\"]\n",
    "                too_short_fastq_file = tech_rep[\"too_short_fastq_file\"]\n",
    "\n",
    "                unique_17_count_df = unique_17_count(tech_rep[\"df\"])\n",
    "\n",
    "                coverage = replicate_coverage(\n",
    "                    lib_df, unique_17_count_df, on_lib=\"unique_17\", on_df=\"unique_17\"\n",
    "                )\n",
    "                fastq_reads = num_of_read_by_fastq(fastq_file)\n",
    "                clipped_fastq_reads = num_of_read_by_fastq(clipped_fastq_file)\n",
    "                unclipped_fastq_reads = num_of_read_by_fastq(unclipped_fastq_file)\n",
    "                too_long_fastq_reads = num_of_read_by_fastq(too_long_fastq_file)\n",
    "                too_short_fastq_reads = num_of_read_by_fastq(too_short_fastq_file)\n",
    "\n",
    "                output_line = (\n",
    "                    f\"{group_name} -> {bio_name} -> {tech_name}:\"\n",
    "                    f\"\\n\\tnumber of aligned reads: {tech_rep['df'].shape[0]} ===> coverage: {coverage}\"\n",
    "                    f\"\\n\\tfastq number of reads: {fastq_reads}\"\n",
    "                    f\"\\n\\tclipped fastq number of reads: {clipped_fastq_reads} ===> clipped reads coverage: {clipped_fastq_reads/fastq_reads}\"\n",
    "                    f\"\\n\\tunclipped fastq number of reads: {unclipped_fastq_reads} ===> unclipped reads coverage: {unclipped_fastq_reads/fastq_reads}\"\n",
    "                    f\"\\n\\ttoo long fastq number of reads: {too_long_fastq_reads} ===> too long reads coverage: {too_long_fastq_reads/fastq_reads}\"\n",
    "                    f\"\\n\\ttoo short fastq number of reads: {too_short_fastq_reads} ===> too short reads coverage: {too_short_fastq_reads/fastq_reads}\"\n",
    "                )\n",
    "\n",
    "                print(output_line)\n",
    "                with open(coverage_file, \"a+\") as out:\n",
    "                    out.write(output_line + \"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "expr_obj = read_experimental_dfs(config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_experiment_coverage(expr_obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6000"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_unique_17s = lib_df[\"unique_17\"]\n",
    "len(all_unique_17s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_experiment_analysis(expr_obj):\n",
    "    expand_output_df = pd.DataFrame(data={\"unique_17\": all_unique_17s})\n",
    "    bio_output_df = pd.DataFrame(data={\"unique_17\": all_unique_17s})\n",
    "\n",
    "    for group in expr_obj[\"groups\"]:\n",
    "        group_name = group[\"name\"]\n",
    "        for bio_rep in group[\"biological_replicates\"]:\n",
    "            bio_name = bio_rep[\"name\"]\n",
    "            bio_count_col_name = f\"{group_name}_{bio_name}\"\n",
    "            bio_count_df = pd.DataFrame(data={\"unique_17\": all_unique_17s, f\"{bio_count_col_name}\": np.zeros(len(all_unique_17s))})\n",
    "            for tech_rep in bio_rep[\"technical_replicates\"]:\n",
    "                tech_name = tech_rep[\"name\"]\n",
    "                \n",
    "                unique_17_count_df = unique_17_count(tech_rep[\"df\"])\n",
    "                tech_count_col_name = f\"{group_name}_{bio_name}_{tech_name}\"\n",
    "                unique_17_count_df = unique_17_count_df.rename(columns={\"count\" : tech_count_col_name})\n",
    "                bio_count_df = bio_count_df.merge(unique_17_count_df, how=\"left\", on=\"unique_17\").fillna(0)\n",
    "                bio_count_df[bio_count_col_name] = bio_count_df[bio_count_col_name] + bio_count_df[tech_count_col_name]\n",
    "            \n",
    "            bio_output_df = bio_output_df.merge(bio_count_df[[\"unique_17\",bio_count_col_name]], how=\"left\", on=\"unique_17\")\n",
    "            expand_output_df = expand_output_df.merge(bio_count_df, how=\"left\", on=\"unique_17\")\n",
    "\n",
    "    return expand_output_df, bio_output_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "expand_output_df, bio_output_df = run_experiment_analysis(expr_obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>unique_17</th>\n",
       "      <th>NCI_R1_bio_1</th>\n",
       "      <th>TREGS_R1_bio_1</th>\n",
       "      <th>NCI_R2_bio_1</th>\n",
       "      <th>TREGS_R2_bio_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACTACCTGAAGAACCTT</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAGCTAAATGGCTGATT</td>\n",
       "      <td>45.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTGACCACACTTACAGT</td>\n",
       "      <td>23.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTGTTGGCGAGCAGTGT</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAATATCGGCGAGCTCT</td>\n",
       "      <td>63.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           unique_17  NCI_R1_bio_1  TREGS_R1_bio_1  NCI_R2_bio_1  \\\n",
       "0  ACTACCTGAAGAACCTT           2.0             0.0           0.0   \n",
       "1  GAGCTAAATGGCTGATT          45.0            12.0          60.0   \n",
       "2  GTGACCACACTTACAGT          23.0            56.0          23.0   \n",
       "3  TTGTTGGCGAGCAGTGT           4.0             9.0           2.0   \n",
       "4  CAATATCGGCGAGCTCT          63.0            28.0          66.0   \n",
       "\n",
       "   TREGS_R2_bio_1  \n",
       "0             0.0  \n",
       "1             5.0  \n",
       "2            45.0  \n",
       "3             3.0  \n",
       "4            23.0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bio_output_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>unique_17</th>\n",
       "      <th>NCI_R1_bio_1</th>\n",
       "      <th>NCI_R1_bio_1_tech_2</th>\n",
       "      <th>NCI_R1_bio_1_tech_3</th>\n",
       "      <th>TREGS_R1_bio_1</th>\n",
       "      <th>TREGS_R1_bio_1_tech_2</th>\n",
       "      <th>TREGS_R1_bio_1_tech_3</th>\n",
       "      <th>NCI_R2_bio_1</th>\n",
       "      <th>NCI_R2_bio_1_tech_2</th>\n",
       "      <th>NCI_R2_bio_1_tech_3</th>\n",
       "      <th>TREGS_R2_bio_1</th>\n",
       "      <th>TREGS_R2_bio_1_tech_2</th>\n",
       "      <th>TREGS_R2_bio_1_tech_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACTACCTGAAGAACCTT</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAGCTAAATGGCTGATT</td>\n",
       "      <td>45.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTGACCACACTTACAGT</td>\n",
       "      <td>23.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTGTTGGCGAGCAGTGT</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAATATCGGCGAGCTCT</td>\n",
       "      <td>63.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           unique_17  NCI_R1_bio_1  NCI_R1_bio_1_tech_2  NCI_R1_bio_1_tech_3  \\\n",
       "0  ACTACCTGAAGAACCTT           2.0                  2.0                  0.0   \n",
       "1  GAGCTAAATGGCTGATT          45.0                 35.0                 10.0   \n",
       "2  GTGACCACACTTACAGT          23.0                 14.0                  9.0   \n",
       "3  TTGTTGGCGAGCAGTGT           4.0                  2.0                  2.0   \n",
       "4  CAATATCGGCGAGCTCT          63.0                 23.0                 40.0   \n",
       "\n",
       "   TREGS_R1_bio_1  TREGS_R1_bio_1_tech_2  TREGS_R1_bio_1_tech_3  NCI_R2_bio_1  \\\n",
       "0             0.0                    0.0                    0.0           0.0   \n",
       "1            12.0                    0.0                   12.0          60.0   \n",
       "2            56.0                   41.0                   15.0          23.0   \n",
       "3             9.0                    0.0                    9.0           2.0   \n",
       "4            28.0                   18.0                   10.0          66.0   \n",
       "\n",
       "   NCI_R2_bio_1_tech_2  NCI_R2_bio_1_tech_3  TREGS_R2_bio_1  \\\n",
       "0                  0.0                  0.0             0.0   \n",
       "1                 41.0                 19.0             5.0   \n",
       "2                 16.0                  7.0            45.0   \n",
       "3                  0.0                  2.0             3.0   \n",
       "4                 20.0                 46.0            23.0   \n",
       "\n",
       "   TREGS_R2_bio_1_tech_2  TREGS_R2_bio_1_tech_3  \n",
       "0                    0.0                    0.0  \n",
       "1                    0.0                    5.0  \n",
       "2                   36.0                    9.0  \n",
       "3                    0.0                    3.0  \n",
       "4                   19.0                    4.0  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expand_output_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "only_tech = expand_output_df.set_index(\"unique_17\").drop([\"NCI_R1_bio_1\",\"TREGS_R1_bio_1\",\"NCI_R2_bio_1\",\"TREGS_R2_bio_1\"], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NCI_R1_bio_1_tech_2</th>\n",
       "      <th>NCI_R1_bio_1_tech_3</th>\n",
       "      <th>TREGS_R1_bio_1_tech_2</th>\n",
       "      <th>TREGS_R1_bio_1_tech_3</th>\n",
       "      <th>NCI_R2_bio_1_tech_2</th>\n",
       "      <th>NCI_R2_bio_1_tech_3</th>\n",
       "      <th>TREGS_R2_bio_1_tech_2</th>\n",
       "      <th>TREGS_R2_bio_1_tech_3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique_17</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACTACCTGAAGAACCTT</th>\n",
       "      <td>1.033924</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAGCTAAATGGCTGATT</th>\n",
       "      <td>18.093662</td>\n",
       "      <td>6.139022</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.461916</td>\n",
       "      <td>24.078922</td>\n",
       "      <td>9.398808</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.225862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGACCACACTTACAGT</th>\n",
       "      <td>7.237465</td>\n",
       "      <td>5.525120</td>\n",
       "      <td>18.952132</td>\n",
       "      <td>9.327395</td>\n",
       "      <td>9.396653</td>\n",
       "      <td>3.462719</td>\n",
       "      <td>18.291279</td>\n",
       "      <td>9.406551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTGGCGAGCAGTGT</th>\n",
       "      <td>1.033924</td>\n",
       "      <td>1.227804</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.596437</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.989348</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.135517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAATATCGGCGAGCTCT</th>\n",
       "      <td>11.890121</td>\n",
       "      <td>24.556087</td>\n",
       "      <td>8.320448</td>\n",
       "      <td>6.218264</td>\n",
       "      <td>11.745816</td>\n",
       "      <td>22.755008</td>\n",
       "      <td>9.653731</td>\n",
       "      <td>4.180689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCCTAGGGGATAGCGAT</th>\n",
       "      <td>9.305312</td>\n",
       "      <td>11.050239</td>\n",
       "      <td>7.395954</td>\n",
       "      <td>1.243653</td>\n",
       "      <td>7.634780</td>\n",
       "      <td>12.861526</td>\n",
       "      <td>9.653731</td>\n",
       "      <td>3.135517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCGAACGTTGTTGCATG</th>\n",
       "      <td>94.604005</td>\n",
       "      <td>82.876795</td>\n",
       "      <td>307.394336</td>\n",
       "      <td>289.149254</td>\n",
       "      <td>82.220711</td>\n",
       "      <td>81.126551</td>\n",
       "      <td>276.401557</td>\n",
       "      <td>224.712055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCTTCGTTTCACTCCTC</th>\n",
       "      <td>3.618732</td>\n",
       "      <td>3.683413</td>\n",
       "      <td>14.329661</td>\n",
       "      <td>11.192874</td>\n",
       "      <td>3.523745</td>\n",
       "      <td>1.978696</td>\n",
       "      <td>9.653731</td>\n",
       "      <td>11.496896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGTTCTAAGTCTTGAAG</th>\n",
       "      <td>1.550885</td>\n",
       "      <td>2.455609</td>\n",
       "      <td>6.933707</td>\n",
       "      <td>12.436527</td>\n",
       "      <td>2.349163</td>\n",
       "      <td>6.430763</td>\n",
       "      <td>8.637549</td>\n",
       "      <td>28.219653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAAGCCACGTACAGACA</th>\n",
       "      <td>5.169618</td>\n",
       "      <td>8.594631</td>\n",
       "      <td>1.386741</td>\n",
       "      <td>18.654791</td>\n",
       "      <td>9.396653</td>\n",
       "      <td>7.420111</td>\n",
       "      <td>1.524273</td>\n",
       "      <td>19.858275</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6000 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   NCI_R1_bio_1_tech_2  NCI_R1_bio_1_tech_3  \\\n",
       "unique_17                                                     \n",
       "ACTACCTGAAGAACCTT             1.033924             0.000000   \n",
       "GAGCTAAATGGCTGATT            18.093662             6.139022   \n",
       "GTGACCACACTTACAGT             7.237465             5.525120   \n",
       "TTGTTGGCGAGCAGTGT             1.033924             1.227804   \n",
       "CAATATCGGCGAGCTCT            11.890121            24.556087   \n",
       "...                                ...                  ...   \n",
       "CCCTAGGGGATAGCGAT             9.305312            11.050239   \n",
       "CCGAACGTTGTTGCATG            94.604005            82.876795   \n",
       "CCTTCGTTTCACTCCTC             3.618732             3.683413   \n",
       "CGTTCTAAGTCTTGAAG             1.550885             2.455609   \n",
       "TAAGCCACGTACAGACA             5.169618             8.594631   \n",
       "\n",
       "                   TREGS_R1_bio_1_tech_2  TREGS_R1_bio_1_tech_3  \\\n",
       "unique_17                                                         \n",
       "ACTACCTGAAGAACCTT               0.000000               0.000000   \n",
       "GAGCTAAATGGCTGATT               0.000000               7.461916   \n",
       "GTGACCACACTTACAGT              18.952132               9.327395   \n",
       "TTGTTGGCGAGCAGTGT               0.000000               5.596437   \n",
       "CAATATCGGCGAGCTCT               8.320448               6.218264   \n",
       "...                                  ...                    ...   \n",
       "CCCTAGGGGATAGCGAT               7.395954               1.243653   \n",
       "CCGAACGTTGTTGCATG             307.394336             289.149254   \n",
       "CCTTCGTTTCACTCCTC              14.329661              11.192874   \n",
       "CGTTCTAAGTCTTGAAG               6.933707              12.436527   \n",
       "TAAGCCACGTACAGACA               1.386741              18.654791   \n",
       "\n",
       "                   NCI_R2_bio_1_tech_2  NCI_R2_bio_1_tech_3  \\\n",
       "unique_17                                                     \n",
       "ACTACCTGAAGAACCTT             0.000000             0.000000   \n",
       "GAGCTAAATGGCTGATT            24.078922             9.398808   \n",
       "GTGACCACACTTACAGT             9.396653             3.462719   \n",
       "TTGTTGGCGAGCAGTGT             0.000000             0.989348   \n",
       "CAATATCGGCGAGCTCT            11.745816            22.755008   \n",
       "...                                ...                  ...   \n",
       "CCCTAGGGGATAGCGAT             7.634780            12.861526   \n",
       "CCGAACGTTGTTGCATG            82.220711            81.126551   \n",
       "CCTTCGTTTCACTCCTC             3.523745             1.978696   \n",
       "CGTTCTAAGTCTTGAAG             2.349163             6.430763   \n",
       "TAAGCCACGTACAGACA             9.396653             7.420111   \n",
       "\n",
       "                   TREGS_R2_bio_1_tech_2  TREGS_R2_bio_1_tech_3  \n",
       "unique_17                                                        \n",
       "ACTACCTGAAGAACCTT               0.000000               0.000000  \n",
       "GAGCTAAATGGCTGATT               0.000000               5.225862  \n",
       "GTGACCACACTTACAGT              18.291279               9.406551  \n",
       "TTGTTGGCGAGCAGTGT               0.000000               3.135517  \n",
       "CAATATCGGCGAGCTCT               9.653731               4.180689  \n",
       "...                                  ...                    ...  \n",
       "CCCTAGGGGATAGCGAT               9.653731               3.135517  \n",
       "CCGAACGTTGTTGCATG             276.401557             224.712055  \n",
       "CCTTCGTTTCACTCCTC               9.653731              11.496896  \n",
       "CGTTCTAAGTCTTGAAG               8.637549              28.219653  \n",
       "TAAGCCACGTACAGACA               1.524273              19.858275  \n",
       "\n",
       "[6000 rows x 8 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reads_per_million_norm(only_tech)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "expand_output_df_csv = os.path.join(output_folder, \"all_replicates_sum.csv\")\n",
    "save_df_to_csv(expand_output_df, expand_output_df_csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "bio_output_df_csv = os.path.join(output_folder, \"all_bio_replicates_sum.csv\")\n",
    "save_df_to_csv(bio_output_df, bio_output_df_csv)"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "70b838359764d1d529f23188ff72e6bf01483479c193c40f5e02842d85becc68"
  },
  "kernelspec": {
   "display_name": "Python 3.8.12 ('bio')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
